### DISCUSSION TITLE
DT misbehaving in several ways even on a brand new deployment

### DISCUSSION THREAD

**Discussion by flemminglau** [DependencyTrack/dependency-track#1875] [Q&A] [answered]
We are running DT 4.5 using docker compose and H2 DB and everything was working perfectly. Even after the upgrade to 4.5. 

Then suddenly our jenkins plugin started to get stuck and never receiving any response. So we added timeouts on the calls but indeed. It just times out.

At the same time we started to see in the UI that the "Audit Vulnerabilities" tab in a Project sometimes never changed from 0 and when opening it the UI just sat there "Loading, Please Wait" forever. Other times the data does show up after a looong time. Nothing is logged while this is happening.

In the log we see instances if the error:
```
2022-08-09 08:00:53,153 ERROR [ServerRuntime$Responder] An I/O error has occurred while writing a response message entity to the container output stream.
org.glassfish.jersey.server.internal.process.MappableException: java.io.IOException: java.util.concurrent.TimeoutException: Idle timeout expired: 30000/30000 ms
        at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:67)
```

So our conclusion was that the H2 DB was somehow corrupt and we had better in any case set up a system using postgresql.

However after setting up the new system we see the exact same behaviour. Same timeouts in Jenkins plugin, same UI behaviour, same error messages in the log.
At least on the new system we are also seeing that the UI gets completely frozen while the Jenkins plugin processing is taking place.

So it seems it is either our configuration or our SBOMs which are causing the issues. Currently we are favoring the latter. 

Any ideas as to how we debug this further?

**Comment by nscuro** [comment]
Can you provide more details on the workload of your instance? 

* How many projects, components does your DT instance manage? 
* How large are the SBOMs you're uploading? How many components do they contain?
* How many resources (CPU cores, RAM) have you allocated to the DT host?

**Reply by flemminglau** [reply]
I can reproduce the issue with a single project of 488 components.
(I.e. after deleting all existing projects it still behaves as described above when loading a single new project via the Jenkins plugin. All processing seems to eventually be done, I get email events and Policy Violations indicated.)

I have 6 CPUs / 16 Gb RAM  on the VM.

Would it make a difference if I clean the DB or is it fine to simply delete all projects?

**Reply by flemminglau** [reply]
I have a hunch that the issues started when I enabled the GHAS connection and I also enabled it on the new system initially. But disabling it now seems to make no difference.

**Reply by nscuro** [reply]
488 components shouldn't be an issue, especially for an instance with those resources allocated to it. I suspected a performance issue but your description raises doubts for me.

Could you post a snippet of your API server logs, starting from when you upload the BOM up the point where all related analyses have been finished?

**Reply by nscuro** [reply]
Are you using a load balancer or reverse proxy that could artificially hold connections? I've never encountered a situation in which DT failed to respond within 5 minutes. The main processing is done asynchronously, so it won't block any requests.

**Reply by flemminglau** [reply]
As mentioned it has been running flawlessly with many similar sized projects in the H2 configuration (which has a smaller HW specs).

And to my surprise the 2 environments behave the same.

I am in a corporate network so have set up proxy settings for DT. But remember it worked before. Jenkins plugin is indeed able to successfully do a "Test Connection". But of course something weird may have changed in my network.

API server log I assume is the log coming out of a "docker logs" call?

```
2022-08-09 13:04:13,486 INFO [MetricsUpdateTask] Completed metrics update on vulnerability database
2022-08-09 13:04:26,967 INFO [BomUploadProcessingTask] Processing CycloneDX BOM uploaded to project: 43baf417-199c-46f3-9e8f-6a445d370be1
2022-08-09 13:04:33,904 INFO [BomUploadProcessingTask] Processing CycloneDX dependency graph for project: 43baf417-199c-46f3-9e8f-6a445d370be1
2022-08-09 13:04:34,306 INFO [BomUploadProcessingTask] Processed 488 components and 0 services uploaded to project 43baf417-199c-46f3-9e8f-6a445d370be1
2022-08-09 13:04:50,702 INFO [InternalAnalysisTask] Starting internal analysis task
2022-08-09 13:04:52,959 INFO [InternalAnalysisTask] Internal analysis complete
2022-08-09 13:04:52,963 WARN [OssIndexAnalysisTask] An API username or token has not been specified for use with OSS Index. Using anonymous access
2022-08-09 13:04:52,964 INFO [OssIndexAnalysisTask] Starting Sonatype OSS Index analysis task
2022-08-09 13:04:56,001 INFO [OssIndexAnalysisTask] Analyzing 80 component(s)
2022-08-09 13:04:57,870 INFO [OssIndexAnalysisTask] Analyzing 92 component(s)
2022-08-09 13:05:00,427 INFO [OssIndexAnalysisTask] Analyzing 99 component(s)
2022-08-09 13:05:02,446 INFO [OssIndexAnalysisTask] Analyzing 100 component(s)
2022-08-09 13:05:04,171 INFO [OssIndexAnalysisTask] Analyzing 88 component(s)
2022-08-09 13:05:04,171 INFO [OssIndexAnalysisTask] Sonatype OSS Index analysis complete
2022-08-09 13:05:04,182 INFO [PolicyEngine] Evaluating 488 component(s) against applicable policies
2022-08-09 13:05:11,937 INFO [PolicyEngine] Policy analysis complete
2022-08-09 13:05:11,938 INFO [MetricsUpdateTask] Executing metrics update for project: 43baf417-199c-46f3-9e8f-6a445d370be1
2022-08-09 13:05:18,176 INFO [MetricsUpdateTask] Completed metrics update for project: 43baf417-199c-46f3-9e8f-6a445d370be1
```

**Reply by nscuro** [reply]
That looks fine so far. It's taking a little long (almost a whole minute) which is surprising. But still nothing that could explain the timeout.

What do you see in your Jenkins build logs? I'm wondering which request is hanging:

* The one in which it uploads the BOM
* One of the polling requests it uses to check whether BOM processing completed

Normally you should see output similar to this:

```
[DependencyTrack] Publishing artifact to Dependency-Track - https://dtrack.example.com/
[DependencyTrack] The artifact was successfully published. You may now navigate to https://dtrack.example.com/projects/ to view the results.
[DependencyTrack] Polling Dependency-Track for BOM processing status
[DependencyTrack] Looking up id of newly created project with name "acme-project" and version "LATEST"
[DependencyTrack] Processing findings
[DependencyTrack] Evaluating new findings against previous build #179
```

**Reply by flemminglau** [reply]
I just tried loading a completely different BOM Downloaded from the old system. I got:

```
2022-08-09 13:11:56,980 WARN [Persist] Insert of object "org.dependencytrack.model.RepositoryMetaComponent@1b4d9af2" using statement "INSERT INTO "REPOSITORY_META_COMPONENT" ("LAST_CHECK","LATEST_VERSION","NAME","NAMESPACE","PUBLISHED","REPOSITORY_TYPE") VALUES (?,?,?,?,?,?)" failed : ERROR: duplicate key value violates unique constraint "REPOSITORY_META_COMPONENT_COMPOUND_IDX"
  Detail: Key ("REPOSITORY_TYPE", "NAMESPACE", "NAME")=(MAVEN, org.aesh, aesh) already exists.
2022-08-09 13:11:56,992 WARN [General] ExecutionContext closed with active transaction, so rolling back the active transaction
2022-08-09 13:11:56,998 ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread
javax.jdo.JDODataStoreException: Insert of object "org.dependencytrack.model.RepositoryMetaComponent@1b4d9af2" using statement "INSERT INTO "REPOSITORY_META_COMPONENT" ("LAST_CHECK","LATEST_VERSION","NAME","NAMESPACE","PUBLISHED","REPOSITORY_TYPE") VALUES (?,?,?,?,?,?)" failed : ERROR: duplicate key value violates unique constraint "REPOSITORY_META_COMPONENT_COMPOUND_IDX"
  Detail: Key ("REPOSITORY_TYPE", "NAMESPACE", "NAME")=(MAVEN, org.aesh, aesh) already exists.
        at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:542)
```

and similar errors in the log.
I never saw errors like that before.

**Reply by nscuro** [reply]
Unfortunately the 

```
Key ("REPOSITORY_TYPE", "NAMESPACE", "NAME")=(MAVEN, org.aesh, aesh) already exists
```

error is somewhat expected when the same component is present multiple times in the same BOM, or multiple BOMs containing the same component are uploaded at the same time. DT will fetch the latest version for all components from a remote repository. This is done concurrently in separate threads, so it may happen that multiple threads try to insert their results at the same time, causing unique constraint violations in the DB. This error doesn't have a negative impact on the system however, so nothing to worry about.

**Reply by flemminglau** [reply]
NEWS:

I (in my desperation) hacked my Jenkins plugin job to upload the same alternative file as above.
That worked. :-)
(The offensive file is currently not in DT at all)

So it seems there is something in my favorite SBOM file which breaks DT. Without it causing any comments in the log.

Something which showed up about a month ago when this started happening. (As the previous 65 uploads had not caused any issues.)

**Reply by nscuro** [reply]
Is there a chance that you can strip all internal information from that BOM (name of your company etc.) and share it with us? Would be truly unfortunate if a BOM could be used to effectively DOS the system.

**Reply by flemminglau** [reply]
Is there a way I can send it to you without making it available to the world?

**Reply by nscuro** [reply]
Sure, if you join the OWASP Slack space, you can send it to me via DM. See https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#asking-questions. My username is `nscur0` over there.

**Reply by flemminglau** [reply]
Today behaviour is different.

I still cannot load the offending file using the Jenkins plugin but I can do it manually with no adverse consequences..
The slowness/failure to show Audit Vulnerabilities data is gone. 
The overall system seems to be much snappier. Almost immediate responses to all navigation.

I did yesterday disable GHAS but yesterday that made no difference. Possibly this is changed today.
On my old H2 based system I see the same change in behaviour and there GHAS is still enabled.

I have attempted to obfuscate the file but possibly the obfuscation removes the issue??? I need to investigate further.

**Reply by flemminglau** [reply]
As it looks now my SBOM file is the culprit but only to the extent that the Jenkins plugin times out when loading it.
The loading completes fine and the result is behaving fine.
As a result also a manual or a curl upload works fine.
 
So it seems the bad behaviour I saw yesterday is unrelated to the fact that the file causes DT to never return a response to the plugin.

With respect to the file I build it using a cyclonedx-cli merge of several files. All of which can all be loaded individually but the merged file causes the timeout. ???

**Reply by flemminglau** [reply]
It seems I need to have an account in OWASP to connect via slack?

**Reply by nscuro** [reply]
You can get an invite to the space [here](https://dependencytrack.org/slack/invite) (as stated in our `CONTRIBUTING.md`) or directly from OWASP at https://owasp.org/slack/invite

**Comment by flemminglau** [comment, marked as answer]
Conclusion:

It turns out that the 2 problems described above are actually unrelated.

The problem of DT not showing "Vulnerability Audit" data magically disappeared overnight from both my systems. So it has not been pursued further. 

Concerning the issue of the Jenkins Plugin timeout it turns out that the root cause is that somehow the response for certain sbom files gets blocked by the enterprise proxy server (which should not really be involved but no_proxy is set up wrongly on the Jenkins server it seems). 
This issue was verified by running the command:
```
curl -v -H 'X-Api-Key: <API-KEY>' https://dtrack.example.com/api/v1/finding/project/<PROJECT_UUID>
``` 
which is the call that the plugin performs and which for this specific file and in our network times out when going via the proxy and otherwise not. 

So I am closing this issue as resolved.
Will raise a separate issue should the slowness ever return.
